### PR TITLE
Put Image first in Tool Either to force Image transform

### LINKED ISSUE #11830: [BUG] TypeError when customizing tools icon using PIL Image
### Software versions

```
Python version      :  3.9.6 | packaged by conda-forge | (default, Jul  6 2021, 08:53:59)
IPython version     :  7.29.0
Tornado version     :  6.1
Bokeh version       :  3.0.0dev1+21.g51a06bb37.dirty
BokehJS static path :  /home/laptop/projects/bokeh/bokeh/server/static
node.js version     :  v16.6.1
npm version         :  7.20.3
```

### Browser name and version

_No response_

### Jupyter notebook / Jupyter Lab version

notebook: 6.4.5

### Expected behavior

Show the figure, along with the added tool, and render the custom icons instead of the default tool icon.

### Observed behavior

Code doesn't run, and raises a `TypeError: Object of type Image is not JSON serializable` more details in the traceback.

### Example code

```Python
from bokeh.plotting import figure, show
from bokeh.models.tools import BoxSelectTool
from PIL import Image
from matplotlib import cm
import numpy as np

myarray = np.array([
    [1, 1, 1, 1], 
    [1, 1, 1, 1], 
    [0, 0, 0, 0],])
im = Image.fromarray(np.uint8(cm.gist_earth(myarray)*255))

x = [1, 2, 3, 4, 5]
y = [6, 7, 2, 4, 5]

# create a new figure
figure = figure(title="",tools="")
figure.add_tools(BoxSelectTool(description="Graph selector", icon=im))

# add a line renderer with legend and line thickness
figure.line(x, y, legend_label="Temp.", line_width=2)

# show the results
show(p)
```

### Stack traceback or browser console output

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
/tmp/ipykernel_962/3797676593.py in <module>
     11 
     12 # show the results
---> 13 show(p)

~/projects/bokeh/bokeh/io/showing.py in show(obj, browser, new, notebook_handle, notebook_url, **kwargs)
    142 
    143     if isinstance(obj, LayoutDOM):
--> 144         return _show_with_state(obj, state, browser, new, notebook_handle=notebook_handle)
    145 
    146     def is_application(obj: Any) -> TypeGuard[Application]:

~/projects/bokeh/bokeh/io/showing.py in _show_with_state(obj, state, browser, new, notebook_handle)
    193 
    194     if state.file or not shown:
--> 195         _show_file_with_state(obj, state, new, controller)
    196 
    197     return comms_handle

~/projects/bokeh/bokeh/io/showing.py in _show_file_with_state(obj, state, new, controller)
    174 
    175     '''
--> 176     filename = save(obj, state=state)
    177     controller.open("file://" + filename, new=NEW_PARAM[new])
    178 

~/projects/bokeh/bokeh/io/saving.py in save(obj, filename, resources, title, template, state)
     96 
     97     filename, resources, title = _get_save_args(state, filename, resources, title)
---> 98     _save_helper(obj, filename, resources, title, template, theme)
     99     return abspath(filename)
    100 

~/projects/bokeh/bokeh/io/saving.py in _save_helper(obj, filename, resources, title, template, theme)
    162     '''
    163     from ..embed import file_html
--> 164     html = file_html(obj, resources, title=title, template=template or FILE, theme=theme)
    165 
    166     with open(filename, mode="w", encoding="utf-8") as f:

~/projects/bokeh/bokeh/embed/standalone.py in file_html(models, resources, title, template, template_variables, theme, suppress_callback_warning, _always_new)
    332 
    333     with OutputDocumentFor(models_seq, apply_theme=theme, always_new=_always_new) as doc:
--> 334         (docs_json, render_items) = standalone_docs_json_and_render_items(models_seq, suppress_callback_warning=suppress_callback_warning)
    335         title = _title_from_models(models_seq, title)
    336         bundle = bundle_for_objs_and_resources([doc], resources)

~/projects/bokeh/bokeh/embed/util.py in standalone_docs_json_and_render_items(models, suppress_callback_warning)
    308     docs_json: Dict[ID, DocJson] = {}
    309     for doc, (docid, _) in docs.items():
--> 310         docs_json[docid] = doc.to_json()
    311 
    312     render_items: List[RenderItem] = []

~/projects/bokeh/bokeh/document/document.py in to_json(self)
    734         # this is a total hack to go via a string, needed because
    735         # our BokehJSONEncoder goes straight to a string.
--> 736         doc_json = self.to_json_string()
    737         return loads(doc_json)
    738 

~/projects/bokeh/bokeh/document/document.py in to_json_string(self, indent)
    766         )
    767 
--> 768         return serialize_json(json, indent=indent)
    769 
    770     def unhold(self) -> None:

~/projects/bokeh/bokeh/core/json_encoder.py in serialize_json(obj, pretty, indent, **kwargs)
    165         indent = 2
    166 
--> 167     return json.dumps(obj, cls=BokehJSONEncoder, allow_nan=False, indent=indent, separators=separators, sort_keys=True, **kwargs)
    168 
    169 

~/miniconda3/envs/bkdev/lib/python3.9/json/__init__.py in dumps(obj, skipkeys, ensure_ascii, check_circular, allow_nan, cls, indent, separators, default, sort_keys, **kw)
    232     if cls is None:
    233         cls = JSONEncoder
--> 234     return cls(
    235         skipkeys=skipkeys, ensure_ascii=ensure_ascii,
    236         check_circular=check_circular, allow_nan=allow_nan, indent=indent,

~/miniconda3/envs/bkdev/lib/python3.9/json/encoder.py in encode(self, o)
    197         # exceptions aren't as detailed.  The list call should be roughly
    198         # equivalent to the PySequence_Fast that ''.join() would do.
--> 199         chunks = self.iterencode(o, _one_shot=True)
    200         if not isinstance(chunks, (list, tuple)):
    201             chunks = list(chunks)

~/miniconda3/envs/bkdev/lib/python3.9/json/encoder.py in iterencode(self, o, _one_shot)
    255                 self.key_separator, self.item_separator, self.sort_keys,
    256                 self.skipkeys, _one_shot)
--> 257         return _iterencode(o, 0)
    258 
    259 def _make_iterencode(markers, _default, _encoder, _indent, _floatstr,

~/projects/bokeh/bokeh/core/json_encoder.py in default(self, obj)
    259             return obj.to_css()
    260         else:
--> 261             return self.transform_python_types(obj)
    262 
    263 #-----------------------------------------------------------------------------

~/projects/bokeh/bokeh/core/json_encoder.py in transform_python_types(self, obj)
    228 
    229         else:
--> 230             return super().default(obj)
    231 
    232     def default(self, obj: Any) -> Any:

~/miniconda3/envs/bkdev/lib/python3.9/json/encoder.py in default(self, o)
    177 
    178         """
--> 179         raise TypeError(f'Object of type {o.__class__.__name__} '
    180                         f'is not JSON serializable')
    181 

TypeError: Object of type Image is not JSON serializable
```

### Screenshots

_No response_

### LINKED ISSUE #12774: [BUG] Can not change icon of a tools
### Software versions

Python version      :  3.9.13 (main, Aug 25 2022, 18:29:29)
IPython version     :  7.31.1
Tornado version     :  6.1
Bokeh version       :  3.0.3
BokehJS static path :  /Users/orenamsalem/opt/anaconda3/lib/python3.9/site-packages/bokeh/server/static
node.js version     :  (not installed)
npm version         :  (not installed)
Operating system    :  macOS-10.16-x86_64-i386-64bit

### Browser name and version

chrome 

### Jupyter notebook / Jupyter Lab version

Jupyter Lab - Version 3.4.4

### Expected behavior

It seems that all options for setting custom icon do not work. 
This simple code preset the problem:
```python
from PIL import Image
from pathlib import Path
import numpy as np 
from bokeh.models import PolyDrawTool
from bokeh.plotting import figure, output_notebook, show
output_notebook()

im_icon = Image.open('B_icon.png')
im_icon = im_icon.convert("RGBA")
imarray_icon  = np.array(im_icon)


# all those do not work
p = figure(title="Simple line example", x_axis_label='x', y_axis_label='y')
# add a line renderer with legend and line thickness to the plot
p.line([1,2], [2,3], legend_label="Temp.", line_width=2)
renderer_in = p.multi_line([], [], line_width=5, alpha=0.4, color='red')


#draw_tool_in = PolyDrawTool(renderers=[renderer_in], num_objects=50,icon='CC.svg')
#draw_tool_in = PolyDrawTool(renderers=[renderer_in], num_objects=50,icon='B_icon.png')
#draw_tool_in = PolyDrawTool(renderers=[renderer_in], num_objects=50,icon='A_icon.icon')
#draw_tool_in = PolyDrawTool(renderers=[renderer_in], num_objects=50,icon=Path('B_icon.png'))
draw_tool_in = PolyDrawTool(renderers=[renderer_in], num_objects=50,icon=imarray_icon)


p.add_tools(draw_tool_in)
show(p)

```
[Here are the files used in this exmaple](https://github.com/bokeh/bokeh/files/10564075/example_icons.zip)

I get the following errors, depending on the attempt:
For rgba numpy array: `Javascript Error: PolyDrawTool(p3720).icon given invalid value: {"0":0,"1":0,"2":0,"3`
For Path('B_icon.png'): `SerializationError: can't serialize <class 'pathlib.PosixPath'>`  (releated #11830 ?)
For all the rest I get:  `Javascript Error: PolyDrawTool(p4389).icon given invalid value: "B_icon.png"`

### Observed behavior

Icon should have change, it did work with with bokeh 2.3 using:
`draw_tool_in.custom_icon = Path('A_icon.png')`

### PR DESCRIPTION
**Pull request by orena1**
Although I am not sure this fix it the best, maybe a better fix will be in Either validation - see [12774](https://github.com/bokeh/bokeh/issues/12774)

- [x] issues: fixes #11830
Also related #11640

### CODE REVIEW

**Comment by bryevdv** [comment]
ref: https://github.com/bokeh/bokeh/issues/12774#issuecomment-1413227669

> I feel like a more general solution here might be to add a transform method to Enum that raises a ValueError for values that are not valid. cc @bokeh/core 

Unless I am mistaken that would allow the transform to be attempted on the subsequent types.

**Comment by mattpap** [comment]
This PR is not a fix, but it could be made semi viable by removing the deprecated `str` case from `Image`'s `validate()` and `transform()`. The proper fix is to make `Either.transform()` to apply type parameter's transform for the first type parameter that was valid (that however isn't that simple, because we allow to disable validation).

**Comment by orena1** [comment]
Hi @mattpap and @bryevdv ,, As I mention a better fix would have been in the Either but was a bit more afraid to change it. I'll be happy to try with your suggestions, but I have a few questions:

@bryevdv 
> Unless I am mistaken that would allow the transform to be attempted on the subsequent types.

Yeah, probably, but as @mattpap said, only if you remove str as a valid option

@mattpap 
> This PR is not a fix

It might not be a good fix, but it does fix the problem, when I make this change in my code it fixes the problem.

In general, I am not sure you want to remove str, as it will remove the ability to supply directly bas64 data (the only workaround that I've found), but I assume you have different reasons to remove str. 

So it means that `Regex(r"^\.")` should be removed from here:
https://github.com/bokeh/bokeh/blob/b9e458cfef6e3d86f79c53b507f08b220bf487ae/src/bokeh/models/tools.py#L173
Right? Can you explain to me a bit the meaning of this `Regex(r"^\.")` , why is it there? I thought that it is for checking that the input is XXX.png, but it does not make sense, as file_paths are  also handled in `Image`

One option will be to apply my current fix for 3.1 and remove str from 3.2 (+apply transform to only types that pass validation) because the text deprecation notice was not very long?

Happy to hear what you think,
Thanks

**Comment by bryevdv** [comment]
> Yeah, probably, but as @mattpap said, only if you remove str as a valid option

@orena1 I don't think that is the case (I think the `str` comment was about this PR) . My suggestion is just a concrete possible way to implement @mattpap general statement:

> The proper fix is to make Either.transform() to apply type parameter's transform for the first type parameter that was valid

It's just that essentially `transform` would be doing some validation, in a sense. As @mattpap mentions, determining and storing the currently valid type at validation time would be ideal, but some users have asked for a method to turn off all validation, which makes that impossible in any simple way. 

FWIW, I think having `Enum.transform` raise a `ValueError` on values that can't be transformed is reasonable.

**Comment by orena1** [comment]
Thanks @bryevdv, it does work, can you have a look at this screen-recording? 
https://www.youtube.com/watch?v=ZR2NkndMhu0
What am I missing? even without removing `str` as valid option

> FWIW, I think having Enum.transform raise a ValueError on values that can't be transformed is reasonable.

Yeah that is good, although what is the transform of `Regex(r"^\.")` it is second in the list, so it's transform will run before.
It will fail for Path() but will run for str. Right?
Thanks

**Comment by bryevdv** [comment]
> Thanks @bryevdv, it does work. 

I think we are talking past each other. I know your change in this PR gets the code working, but it's not especially robust to have to rely on a precise ordering in this way. 

> Yeah that is good, although what is the transform of Regex(r"^\.")

I had actually somehow missed the `Regex(r"^\.")` I was there at all.  am not sure what it is there for, to be honest.

**Comment by orena1** [comment]
Ok, I think that might be some of the reasons that I did not understand. :) 
So Regex will accept a string, so its validate will also work, and therefore (as it is second) its transform will work even in the case where we do the transform of what was validated. Right?
Anyway, let's see maybe @mattpap knows what the Regex is doing there.
Thanks!

**Comment by bryevdv** [comment]
@mattpap can you comment on the `Regex(r"^.")` part? I don't recall why it is there or necessary.

**Comment by mattpap** [comment]
> can you comment on the `Regex(r"^.")` part? I don't recall why it is there or necessary.

This is actually not up to date, because bokehjs supports the following:
```ts
Nullable(Or(ToolIcon, Regex(/^--/), Regex(/^\./), Regex(/^data:image/))
```
The regex allows for specifying a CSS class that defines an icon. This isn't currently that useful with shadow DOM in place, because it isn't possible to attach a stylesheet to a tool's button. One would have to construct a tool button manually and add it to a toolbar, but that API wasn't made public (to bokeh) yet.

**Comment by orena1** [comment]
Got it, I think I fix the pull request accordingly

**Comment by bryevdv** [comment]
@orena1 I think for expediency we could merge this for 3.1 with the quick fix, but can you do the following:

* add a comment above the change noting that the order is significant
* open a separate issue regarding  
  > make Either.transform() to apply type parameter's transform for the first type parameter that was valid

**Comment by orena1** [comment]
Thanks @bryevdv, I've added a comment as you suggested and also created a better fix in this pull request https://github.com/bokeh/bokeh/pull/12815

**Review by bryevdv — APPROVED** [review]

File: src/bokeh/models/tools.py [hunk 1]
```diff
@@ -169,8 +169,9 @@ class Tool(Model):
     # explicit __init__ to support Init signatures
     def __init__(self, *args, **kwargs) -> None:
         super().__init__(*args, **kwargs)
-
-    icon = Nullable(Either(Enum(ToolIcon), Regex(r"^\."), Image), help="""
+
```
Comment (bryevdv) [inline src/bokeh/models/tools.py]:
```suggestion

```

**Comment by bryevdv** [comment]
Thanks for the PR @orena1 I pushed a trivial linter fix so that I can merge this tonight for upcoming 3.1. We can continue the other broader fix PR on `branch-3.2`

**Comment by bryevdv** [comment]
macos job failures unrelated, FYI @bokeh/core I have opened a GH support ticket regarding the OSX runner issues

**Comment by orena1** [comment]
Great!

**Pull request merged**